File: internal/ops/nanoid.go
package ops

import (
	"fmt"
	"io"
	"strings"
)

const (
	repoIDLength    = 6
	repoIDAlphabet  = "useandom26T198340PX75pxJACKVERYMINDBUSHWOLFGQZbfghjklqvwyzrict"
	sessionIDLength = 6
)

func generateNanoID(reader io.Reader, length int) (string, error) {
	const mask = byte(63)
	if length <= 0 {
		return "", fmt.Errorf("nanoid length must be positive")
	}
	out := make([]byte, 0, length)
	buf := make([]byte, length)
	for len(out) < length {
		need := length - len(out)
		if need > len(buf) {
			need = len(buf)
		}
		if _, err := io.ReadFull(reader, buf[:need]); err != nil {
			return "", err
		}
		for _, b := range buf[:need] {
			idx := int(b & mask)
			if idx >= len(repoIDAlphabet) {
				continue
			}
			out = append(out, repoIDAlphabet[idx])
			if len(out) == length {
				break
			}
		}
	}
	return string(out), nil
}

func validRepoID(value string) bool {
	if len(value) != repoIDLength {
		return false
	}
	for i := 0; i < len(value); i++ {
		if strings.IndexByte(repoIDAlphabet, value[i]) == -1 {
			return false
		}
	}
	return true
}

File: src/ops/nanoid.rs
use crate::{LooprError, LooprResult};

const REPO_ID_LENGTH: usize = 6;
const REPO_ID_ALPHABET: &str = "useandom26T198340PX75pxJACKVERYMINDBUSHWOLFGQZbfghjklqvwyzrict";

pub trait RandomSource {
    fn fill(&mut self, buf: &mut [u8]) -> LooprResult<()>;
}

pub struct OsRandom;

impl RandomSource for OsRandom {
    fn fill(&mut self, buf: &mut [u8]) -> LooprResult<()> {
        getrandom::fill(buf).map_err(|err| LooprError::new(format!("random bytes: {}", err)))?;
        Ok(())
    }
}

pub fn generate_nanoid(source: &mut dyn RandomSource, length: usize) -> LooprResult<String> {
    if length == 0 {
        return Err(LooprError::new("nanoid length must be positive"));
    }
    let alphabet = REPO_ID_ALPHABET.as_bytes();
    let mut output = String::with_capacity(length);
    let mut buf = vec![0u8; length];
    while output.len() < length {
        let needed = length - output.len();
        let take = needed.min(buf.len());
        source.fill(&mut buf[..take])?;
        for byte in &buf[..take] {
            let idx = (byte & 63) as usize;
            if idx >= alphabet.len() {
                continue;
            }
            output.push(alphabet[idx] as char);
            if output.len() == length {
                break;
            }
        }
    }
    Ok(output)
}

pub fn repo_id_length() -> usize {
    REPO_ID_LENGTH
}

pub fn repo_id_alphabet() -> &'static str {
    REPO_ID_ALPHABET
}
